### ISSUE TITLE
invokes on partially configured provider should error instead of returning empty response

### ISSUE THREAD

**Issue by viveklak** [pulumi/pulumi#10209]
> Granted an empty yaml body should be handled better here of course, but it seems unfortunate that we can't invoke the yaml decode function in previews (thus not be able to provide a preview of the resources to be created by loading the ConfigFile) if the provider has unknowns, since in this case its entirely feasible to perform the invoke even if the underlying kubernetes cluster is unavailable. At the very least, it would be good to allow the provider to handle this situation (preview/unconfigured provider) and not short circuit the request entirely within the engine.

So either we should only support Output based invokes, or we should only allow immediate values to provider construction. Because otherwise you end up in this state where we don't have a provider yet because of unknowns but you can start an invoke with a non-output response which can't indicate unknownness.

Short term we could error out the invoke if the provider was unknown? That's probably a better error than returning undef.

_Originally posted by @Frassle in https://github.com/pulumi/pulumi-kubernetes/issues/2038#issuecomment-1169710008_

**Comment by heyzling** [comment]
Some cases demand dynamic provider configuration. Example case: create Kubernetes and deploy some infrastructure addons to it. In my case it looks like this: create Kubernetes, get kubeconfig, install Ingress, GAP stack, filebeat, etc.. Kubeconfig is dynamic Output here, and often contains short-lived tokens for authentication. Hardcoding kubeconfig is not very convenient option, because it demands using some hacky workarounds outside of Pulumi code scope.

**Comment by thomas11** [comment]
related: pulumi/pulumi-digitalocean#341

**Comment by EronWright** [comment]
I would advocate for a solution that allows the provider to produce unknowns when the invoke call cannot be made.  Raising an error might be good basis for that. In the meantime, the provider must detect whether the result is the zero value, and synthesize an 'unknown' value.

**Comment by Frassle** [comment]
> I would advocate for a solution that allows the provider to produce unknowns when the invoke call cannot be made.

There's two issues with that. First is that invokes don't always use outputs, some invokes are just promise typed and that can't represent `unknown`. So the only reasonable thing to do for those invokes is error.
Secondly there's currently no way for a provider to tell the engine "it's ok to call Configure on me with unknowns" because currently Configure is the method that returns all the feature flags of what the provider supports to the engine. We'd need to add a new PreConfigure command for the provider to have the chance to tell the engine it understands unknowns on Configure.

**Comment by EronWright** [comment]
@Frassle I agree that the invoke itself should error. I'm advocating for having the provider catch the error and then synthesize unknown values, as shown in the example.  The alternative it to let the outputs be uninitialized, which causes various downstream issues.

**Comment by Frassle** [comment]
> I'm advocating for having the provider catch the error and then synthesize unknown values, as shown in the example. 

Your example is pretty unique in that these are overlay functions added to the provider SDK. Under normal SDK generation there's currently no way we can codegen this in general.

> This isn't entirely possible today, e.g. because some SDKs cannot create unknown values. 

Yeh we should probably expose this everywhere just to be consistent and its often useful in tests.

**Comment by EronWright** [comment]
@Frassle is there codegen for invokes? I would guess that most invokes are made by hand-written provider SDK code, typically (but not exclusively) for overlay resources.

To reiterate my suggestion, it would be:
1. add a flag to InvokeOptions called `FailOnPreview` or similar, to opt-in to new behavior.
2. New behavior, in preview mode, throw a specific error upon invoke.
3. For each Pulumi SDK, expose the ability to create an unknown value from scratch.
4. Update the providers to use the new flag, catch the error, and synthesize unknowns as necessary.
